### PR TITLE
feat: Support top-p and top-k

### LINKED ISSUE #69: Top-p/Top-k Sampling Params handling in VLLM v1

### PR DESCRIPTION
**Pull request by zhandaz**
# What does this PR do ?

This PR implements top-k and top-p sampling in training to to ensure consistency between training and inference with sampling parameters.

1. **Why top-p and top-k are a bit tricky?** In TP, the vocab is sharded across GPUs (vocab parallel). However, top-p and top-k require the entire probabilities across the full vocabulary, not individual shards. This means applying all-gather to get the full logits, which leads to communication overhead and the large memory consumption.
2. **Solution**: We convert from vocab-parallel to batch-sequence-parallel layout via `all-to-all` communication, apply filtering on the full vocabulary, then convert back. 

### Detailed Changes

1. In `nemo_rl/model/policy/utils.py`: added the `TrainingSamplingParams` and implemented `apply_top_k_top_p` with proper autograd support.
2. In `nemo_rl/distributed/model_utils.py`: implemented `DistributedLogprobWithSampling` and `ChunkedDistributedLogprobWithSampling`, and integrated into all logprob computation paths (TP, CP, DTensor, packed sequences).
3. Loss Functions and policy workers: updated `ClippedPGLossFn`, `NLLLoss`, and `DPOLossFn` to accept `sampling_params`; changed the policy workers call the functions with `sampling_params`.

### Tests

- `test_top_k_top_p_filtering_forward_backward`: Tests top_k and top_p functionalities. Test gradient masking correctness.
- `test_distributed_logprob_with_sampling`: Tests TP distributed implementation.
- `test_vllm_policy_logprob_agreement_with_sampling`: Test that policy worker logprobs match vLLM with sampling parameters.

We have also tested that with `temperature != 1`, the logprobs match. See `test_vllm_policy_logprob_agreement_with_sampling`.

# Issues
List issues that this PR closes ([syntax](https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)):

# Usage
* **You can potentially add a usage example below**

```python
# Add a code snippet demonstrating how to use this
```

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](/NVIDIA-NeMo/RL/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA-NeMo/RL/blob/main/docs/testing.md) for how to run tests
- [ ] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA-NeMo/RL/blob/main/docs/documentation.md) for how to write, build and test the docs.

# Additional Information
* ...

## Release Notes

* **New Features**
  * Enhanced generation with top-k/top-p filtering support for improved sampling control
  * Improved temperature scaling during model generation
  * Better distributed sampling computation paths

* **Bug Fixes**
  * Fixed NaN issues in gradient computations during sampling operations
  * Adjusted sampling parameter validation thresholds for better compatibility

* **Tests**
  * Added comprehensive tests for sampling functionality and distributed computation paths

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: nemo_rl/algorithms/loss_functions.py [hunk 1]
```diff
@@ -240,17 +247,27 @@ def __call__(
...
         else:
             next_token_logits_wo_last = next_token_logits[
                 :, :-1
             ]  # Remove last position's logits
+            # Apply top-k and top-p filtering
+            next_token_logits_wo_last, _ = apply_top_k_top_p(
+                next_token_logits_wo_last,
+                top_k=sampling_params.top_k if sampling_params is not None else None,
+                top_p=sampling_params.top_p if sampling_params is not None else None,
+            )
```
[the remark this answers is not in the document]
Comment (zhandaz) [reply nemo_rl/algorithms/loss_functions.py]: Fixed

File: nemo_rl/distributed/model_utils.py [hunk 2]
```diff
@@ -255,6 +262,352 @@ def backward(
...
+
+        # Apply top-k and top-p filtering locally (returns keep_mask for gradient)
+        logits, keep_mask = apply_top_k_top_p(
+            seq_parallel_logits, top_k=top_k, top_p=top_p
+        )
+
+        # Compute log softmax
+        log_probs = torch.nn.functional.log_softmax(
+            logits.to(dtype=torch.float32), dim=-1
+        )
```
[the remark this answers is not in the document]
Comment (zhandaz) [reply nemo_rl/distributed/model_utils.py:328]: No. The temperature is scaled inside the policy workers. According to vllm's implementation, the temperature is applied before top-p/k. Since it is an element-wise scaling, it is not related to distributed and can be applied in the policy workers directly.

File: nemo_rl/models/policy/utils.py [hunk 3]
```diff
@@ -123,30 +200,66 @@ def apply_top_k_top_p(
...
 
     # Re-sort the probabilities
     logits = logits_sort.scatter(dim=-1, index=logits_idx, src=logits_sort)
-    return logits
+    if top_k_keep_mask is not None:
+        keep_mask = torch.logical_and(top_k_keep_mask, top_p_keep_mask)
+    else:
+        keep_mask = top_p_keep_mask
+
+    return logits, keep_mask
```
[the remark this answers is not in the document]
Comment (zhandaz) [reply nemo_rl/models/policy/utils.py]: Good catch! Fixed. Will push the commit later.

**Review by terrykong — COMMENTED** [review]
awesome work @zhandaz !

@gshennvm can you review?

@zhandaz can you include convergence plots? maybe at least small qwen out to a few hundred steps to show no collapse and low gen_kl_error? i think we should have at least one nightly test that tests T!=1 & non trivial top-p | top-k, maybe a variant of the qwen30b recipe with just those things changed so that it's comparable?

**Comment by yuki-97** [comment]
closes since https://github.com/NVIDIA-NeMo/RL/pull/2053 merged.
